conn.go

```diff
@@ -975,29 +936,25 @@ func (c *conn) saslOutcome() (stateFunc, error) {
 	return c.negotiateProto, nil
 }
 
-// readFrame is used during connection establishment to read a single frame.
+// readSingleFrame is used during connection establishment to read a single frame.
 //
-// After setup, conn.mux handles incoming frames.
-// used externally by SASL only.
-func (c *conn) readFrame() (frames.Frame, error) {
-	var deadline <-chan time.Time
+// After setup, conn.connReader handles incoming frames.
+func (c *conn) readSingleFrame() (frames.Frame, error) {
 	if c.connectTimeout != 0 {
-		deadline = time.After(c.connectTimeout)
+		_ = c.net.SetDeadline(time.Now().Add(c.connectTimeout))
+		defer func() { _ = c.net.SetDeadline(time.Time{}) }()
 	}
 
-	var fr frames.Frame
-	select {
-	case fr = <-c.rxFrame:
-		return fr, nil
-	case err := <-c.connErr:
-		return fr, err
-	case p := <-c.rxProto:
-		return fr, fmt.Errorf("unexpected protocol header %#v", p)
-	case <-deadline:
-		return fr, errors.New("amqp: timeout waiting for response")
+	fr, err := c.readFrame()
+	if err != nil {
+		return frames.Frame{}, err
 	}
+
+	return fr, nil
 }
 
+const protoHeaderSize = 8
```

Comment:
NIT: Looks like you can just define this inside of readProtoHeader() and limit its scope a bit.

Reply:
fixed